test/integration/backward_compatible/ListenersOnReconnectTest.js

```diff
@@ -86,69 +95,83 @@ describe('ListenersOnReconnectTest', function () {
-                });
-                map = await client.getMap('testmap');
-
-                const listener = {
-                    added: (entryEvent) => {
-                        try {
-                            expect(entryEvent.name).to.equal('testmap');
-                            expect(entryEvent.key).to.equal('keyx');
-                            expect(entryEvent.value).to.equal('valx');
-                            expect(entryEvent.oldValue).to.be.equal(null);
-                            expect(entryEvent.mergingValue).to.be.equal(null);
-                            expect(entryEvent.member).to.not.be.equal(null);
-                            done();
-                        } catch (err) {
-                            done(err);
-                        }
+        it('restart member, listener still receives map.put event [smart=' + isSmart + ']', async function () {
+            const deferred = Util.deferredPromise();
+            const member = await RC.startMember(cluster.id);
+            client = await Client.newHazelcastClient({
+                clusterName: cluster.id,
+                network: {
+                    smartRouting: isSmart
+                }
+            });
+            map = await client.getMap('testmap');
+
+            const listener = {
+                added: (entryEvent) => {
+                    try {
+                        expect(entryEvent.name).to.equal('testmap');
+                        expect(entryEvent.key).to.equal('keyx');
+                        expect(entryEvent.value).to.equal('valx');
+                        expect(entryEvent.oldValue).to.be.equal(null);
+                        expect(entryEvent.mergingValue).to.be.equal(null);
+                        expect(entryEvent.member).to.not.be.equal(null);
+                        deferred.resolve();
+                    } catch (err) {
+                        deferred.reject(err);
                     }
-                };
-                await map.addEntryListener(listener, 'keyx', true);
-
-                await RC.terminateMember(cluster.id, member.uuid);
-                await RC.startMember(cluster.id);
-
-                await TestUtil.promiseWaitMilliseconds(8000);
-                return map.put('keyx', 'valx');
-            }
-
-            testListener(done).catch(done);
+                }
+            };
+            const registrationId = await map.addEntryListener(listener, 'keyx', true);
+
+            await RC.terminateMember(cluster.id, member.uuid);
+            // Assert that the connection is closed and the listener is removed.
+            await TestUtil.assertTrueEventually(async () => {
+                const activeConnections = TestUtil.getConnections(client);
+                expect(activeConnections.length).to.be.equal(0);
+
+                const activeRegistrations = TestUtil.getActiveRegistrations(client, registrationId);
+                expect(activeRegistrations.size).to.be.equal(0);
```

Comment:
While reviewing this PR, I understand that this code in the master will fail in backward compatibility tests :sweat_smile: 

For example, the 5.0 client will not have the code that removes registration when the connection is closed. So, this assertion will fail for them. We need to fix the code in the master

Reply:
🤔 Right, what to do then?